ilastik/workflows/carving/carvingGui.py

```diff
@@ -277,38 +275,41 @@ def validate():
 
     def onSaveButton(self):
         logger.info("save object as?")
-        if self.topLevelOperatorView.dataIsStorable():
-            prevName = ""
-            if self.topLevelOperatorView.hasCurrentObject():
-                prevName = self.topLevelOperatorView.getCurrentObjectName()
-            if prevName == "<not saved yet>":
-                prevName = ""
-            name = self.saveAsDialog(name=prevName)
-            if name is None:
-                return
-            namesInUse = self.getObjectNames()
-            if name in namesInUse and name != prevName:
-                QMessageBox.critical(
-                    self,
-                    "Save Object As",
-                    "An object with name '%s' already exists.\nPlease choose a different name." % name,
-                )
-                return
-            self.topLevelOperatorView.saveObjectAs(name)
-            logger.info("save object as %s" % name)
-            if prevName != name and prevName != "":
-                self.topLevelOperatorView.deleteObject(prevName)
-            elif prevName == name:
-                self._renderMgr.removeObject(prevName)
-                self._renderMgr.invalidateObject(prevName)
-                self._shownObjects3D.pop(prevName, None)
-        else:
+        if not self.topLevelOperatorView.CanObjectBeSaved.value:
             msgBox = QMessageBox(self)
             msgBox.setText("The data does not seem fit to be stored.")
             msgBox.setWindowTitle("Problem with Data")
             msgBox.setIcon(2)
             msgBox.exec_()
             logger.error("object not saved due to faulty data.")
+            return
+
+        old_name = self.topLevelOperatorView.getCurrentObjectName()
+        saved_object_names = self.getObjectNames()
+        was_object_previously_saved = old_name in saved_object_names
+
+        new_name = self.saveAsDialog(name=(old_name if was_object_previously_saved else ""))
+        if new_name is None:
+            return
+
+        is_name_changed = new_name != old_name
+        if is_name_changed and new_name in saved_object_names:
```

Comment:
It seems impossible to have `new_name in saved_object_names` because `saveAsDialog` doesn't let you to proceed if the name one entered is in `getObjectNames()`. However, it may be possible that `validate()` in `saveAsDialog()` is not triggered if `QLineEdit.textChanged` signal is not emitted, which is possible if one never changes text in line input (I'm not sure whether it's true or not...)

Reply:
Nice catch. The text input is pre-filled with the old name (if there was one), so to obtain a name from another already saved object, the user has to change the text. And I think we should be able to trust QT to ensure the signal gets emitted.

Would you prefer the lines thrown out in that case?

Reply:
Hmm, I think the dialog function as least needs to call `validate()` itself in order to set up GUI state properly. Maybe also receive `namesInUse` as a parameter, and don't generate a unique object name by itself.